### PR TITLE
[Segment Replication] [Remote Store] Replace overriding mockInternalEngine() in test classes with NRTReplicationEngine.

### LINKED ISSUE #11715: [Segment Replication] [Remote Store] Replace overriding mockInternalEngine() in test classes with NRTReplicationEngine
### Is your feature request related to a problem? Please describe

Currently many test classes related to segment replication and remote store override [mockInternalEngine()](https://github.com/opensearch-project/OpenSearch/blob/7b1c2c78c16d566da93001d0f6a8224b2de9e8d0/test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java#L2169) to skip usage of Internal Engine for replica shards. Instead of overriding this in every test class we can conditionally return NRTReplicationEngine() from [MockEngineFactory](https://github.com/opensearch-project/OpenSearch/blob/7b1c2c78c16d566da93001d0f6a8224b2de9e8d0/test/framework/src/main/java/org/opensearch/test/engine/MockEngineFactory.java#L48)

### Describe the solution you'd like

Conditionally return NRTReplicationEngine() from [MockEngineFactory](https://github.com/opensearch-project/OpenSearch/blob/7b1c2c78c16d566da93001d0f6a8224b2de9e8d0/test/framework/src/main/java/org/opensearch/test/engine/MockEngineFactory.java#L48)

### Related component

Indexing:Replication

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Rishikesh1159**
### Description
This PR Conditionally(When segment replication is enabled and shard is replica) returns NRTReplicationEngine() from [MockEngineFactory](https://github.com/opensearch-project/OpenSearch/blob/7b1c2c78c16d566da93001d0f6a8224b2de9e8d0/test/framework/src/main/java/org/opensearch/test/engine/MockEngineFactory.java#L48)

### Related Issues
Resolves #11715 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by mch2 — APPROVED** [review]
This LGTM overall.  For the tests that you are removing the override, can you run pls them a few times for sanity? To this point they have never had a mock IE and don't want to add flakiness.

**Review by Poojita-Raj — APPROVED** [review]

File: test/framework/src/main/java/org/opensearch/test/engine/MockEngineFactory.java [hunk 1]
```diff
@@ -46,6 +47,6 @@ public MockEngineFactory(Class<? extends FilterDirectoryReader> wrapper) {
 
     @Override
     public Engine newReadWriteEngine(EngineConfig config) {
```
Comment (Poojita-Raj) [inline test/framework/src/main/java/org/opensearch/test/engine/MockEngineFactory.java:49]: A comment explaining why this change is needed could be useful for contributors without segrep knowledge.

**Review by kotwanikunal — APPROVED** [review]

File: test/framework/src/main/java/org/opensearch/test/engine/MockEngineFactory.java (hunk 1 above)
Comment (Rishikesh1159) [reply test/framework/src/main/java/org/opensearch/test/engine/MockEngineFactory.java:49]: Sure

File: test/framework/src/main/java/org/opensearch/test/engine/MockEngineFactory.java [hunk 2]
```diff
@@ -46,6 +47,8 @@ public MockEngineFactory(Class<? extends FilterDirectoryReader> wrapper) {
...
-        return new MockInternalEngine(config, wrapper);
+
+        // When Segment Replication is enabled on an index replica shards use NRTReplicationEngine instead of InternalEngine.
```
Comment (andrross) [inline test/framework/src/main/java/org/opensearch/test/engine/MockEngineFactory.java:51]: This comment explains "what" is happening (which is apparent from reading the code), but not so much "why". I'd recommend a comment more like:

> Segment replication replicas (i.e. read only replicas) do not use an InternalEngine so a MockInternalEngine will not work and an NRTReplicationEngine must be used instead. The primary shards for these indexes will still use a MockInternalEngine.

Comment (Rishikesh1159) [reply test/framework/src/main/java/org/opensearch/test/engine/MockEngineFactory.java:51]: Makes sense. Will update it. Thank you!

**Review by andrross — APPROVED** [review]
